### PR TITLE
dxf: merge loops in Run and RunStep of task executor

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #57497

Problem Summary:

### What changed and how does it work?
- when detecting whether params are modified, we need to make sure new subtask started after we finished `modify` can use the latest param, especially task concurrency which need to alloc slots from slot manager, and it might require recreating the task executor, if we want avoid create/init step executor when we need recreate the task executor, we have to handle param modify detection on both loop, so we merge them to avoid handle them twice.
- remove `HasSubtasksInStates` which is not needed now, add `GetStep` to step executor
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/disttask/framework/taskexecutor/task_executor_test.go [hunk 1]
```diff
@@ -459,78 +692,10 @@ func TestCheckBalanceSubtask(t *testing.T) {
 	})
 }
 
-func TestExecutorErrHandling(t *testing.T) {
```
Comment (D3Hunter) [inline pkg/disttask/framework/taskexecutor/task_executor_test.go:462]: some are duplicate with TestExecutorRun, some moved to TestExecutorRun

**Comment by D3Hunter** [comment]
@lance6716 3/4 changes are tests

**Comment by D3Hunter** [comment]
@lance6716

File: pkg/disttask/framework/taskexecutor/task_executor.go [hunk 2]
```diff
@@ -230,181 +251,166 @@ func (e *BaseTaskExecutor) Run(resource *proto.StepResource) {
...
-					return err
-				}
-				return ErrNonIdempotentSubtask
-			}
-			e.logger.Info("subtask in running state and is idempotent",
+	return nil
+}
+
+func (e *BaseTaskExecutor) cleanStepExecutor() {
+	if err2 := e.stepExec.Cleanup(e.ctx); err2 != nil {
```
Comment (lance6716) [inline pkg/disttask/framework/taskexecutor/task_executor.go:320]: check `e.stepExec` and `Cleanup` is nil? Not sure it will happen

**Review by lance6716 — APPROVED** [review]

File: pkg/disttask/framework/taskexecutor/task_executor.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/disttask/framework/taskexecutor/task_executor.go:320]: `e.stepExec != nil` is checked at caller

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**